contracts/src/WarpProtocolRegistry.sol

```diff
@@ -0,0 +1,195 @@
+     */
+    function getVersionFromAddress(
+        address protocolAddress
+    ) external view returns (uint256) {
+        return _addressToVersion[protocolAddress];
+    }
+
+    /**
+     * @dev Gets the latest protocol version.
+     */
+    function getLatestVersion() external view returns (uint256) {
+        return _latestVersion;
+    }
+
+    /**
+     * @dev Gets and verifies for a warp out-of-band message, and adds the new protocol version
+     * addres to the registry.
+     */
+    function _addProtocolVersion(uint32 messageIndex) internal virtual {
+        // Get and validate for a warp out-of-band message.
+        (WarpMessage memory message, bool valid) = WARP_MESSENGER
+            .getVerifiedWarpMessage(messageIndex);
+        if (!valid) {
+            revert InvalidWarpMessage();
+        }
+        if (message.sourceChainID != _chainID) {
+            revert InvalidSourceChainID();
+        }
+        if (message.originSenderAddress != VALIDATORS_SOURCE_ADDRESS) {
+            revert InvalidOriginSenderAddress();
+        }
+        if (message.destinationChainID != _chainID) {
+            revert InvalidDestinationChainID();
+        }
+        if (message.destinationAddress != address(this)) {
+            revert InvalidDestinationAddress();
+        }
+
+        ProtocolRegistryEntry memory entry = abi.decode(
+            message.payload,
+            (ProtocolRegistryEntry)
+        );
+
+        _addToRegistry(entry);
+    }
+
+    /**
+     * @dev Adds the new protocol version address to the registry.
+     *
+     * Emits a {AddProtocolVersion} event when successful.
+     * Note: `protocolAddress` doesn't have to be a contract address, this is primarily
+     * to support the case we want to register a new protocol address meant for a security patch
+     * before the contract is deployed, to prevent the vulnerabilitiy from being exposed before registry update.
+     * Requirements:
+     *
+     * - `version` must be the increment of the latest version.
+     */
+    function _addToRegistry(
+        ProtocolRegistryEntry memory entry
+    ) internal virtual {
+        // Check that the version is the increment of the latest version.
+        if (entry.version != _latestVersion + 1) {
+            revert InvalidProtocolVersion();
+        }
+
+        if (entry.protocolAddress == address(0)) {
+            revert InvalidProtocolAddress();
+        }
+
+        _latestVersion++;
+        _versionToAddress[entry.version] = entry.protocolAddress;
+        _addressToVersion[entry.protocolAddress] = entry.version;
+        emit AddProtocolVersion(entry.version, entry.protocolAddress);
+    }
+
+    /**
+     * @dev Gets the address of a protocol version.
+     * Requirements:
+     *
+     * - `version` must be a valid version, i.e. greater than 0 and not greater than the latest version.
+     */
+    function _getAddressFromVersion(
```

Comment:
This seems like it can be an external view function

Reply:
there is a correspnoding `external` function, but we also have this `internal` function so that derived contracts can use it, for example in `TeleporterRegistry`

Reply:
Do we need both functions? The `external` function just immediately calls this one.

Reply:
We need it in the derived contracts like `TeleporterRegistry`, so that the derived contracts don't need to call `this.getAddressFromVersion` which costs more gas than using internal `_getAddressFromVersion()`

Reply:
Is there a significant gas savings to this approach as opposed to making `getAddressFromVersion` and `getVersionFromAddress` both be `public`? My understanding is that `public` is essentially the same as `external + internal`, which is the behavior we want here.

Reply:
Haven't tested the gas and whether it's significant, but would guess it's nothing significant. The `internal` also allows derived contracts to override, but added a new issue, where it should also revisit the `external` + `internal` vs `public`. 

Any reason for opposing external + internal? Maybe it complicates the contract more, or not sure if it's in regards to increasing contract size?

Reply:
Could it be `public virtual` to allow for inheriting contracts to override it still?

And no strong reason, but it increase the size of the contract, and increase the mental load on someone trying to understand the contract (they have to remember or reason about why the function is duplicated in the way that it is).